### PR TITLE
[sw, rom_ext_signer] Add first unittest

### PR DESCRIPTION
**Pull request by silvestrst**
# Introduction

Rust + Cargo have functionality for creating and running unit and integration tests. This change introduces a single unittest for one function (eventually we will add more) - **it would be great to gather some feedback on the approach and style of this test**.

## Rust unittests

- `use super::*;` inside `mod tests {` brings all the dependencies and types that are available in the parent into the scope, it also allows to access private structure fields, like `data` in the code snippet below:
```
/// Stripped binary image buffer.                                                                                                                                                              
pub struct RawImage {
    data: Vec<u8>,
    path: PathBuf,
}
```

## Cargo

To run unittests directly - you would have to navigate to the workspace directory and run `Cargo test --workspace` (`--workspace` flag means that the tests for all of the workspace modules will be built and run). However, the way test is implemented - it uses `MESON_SOURCE_ROOT` environment variable that is set via `meson -> invoke_cargo.sh` script. The alternative would be to instead of `MESON_SOURCE_ROOT` - to use `Cargo.toml` relative directory.

# Please read the individual commit descriptions for more detailed information

### CODE REVIEW

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 1]
```diff
@@ -184,19 +184,102 @@ impl RawImage {
...
+                    offset: String::from("0x77777777"),
+                    checksum: String::from("0x77777777"),
+                },
+            ],
+        };
+        let mut image = RawImage::new(&config.input_files.image_path);
+
+        // Make sure that the image is blank to start with.
+        for byte in image.data.iter() {
+            assert_eq!(*byte, 0);
```
Comment (mcy) [inline sw/host/rom_ext_image_tools/signer/image/src/image.rs:269]: Perhaps worth including the index in the assert message?

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 2]
```diff
@@ -184,19 +184,102 @@ impl RawImage {
...
+
+        let check_u32 = |offset, value_as_str| {
+            let input_val = str_to_u64(value_as_str) as u32;
+            let read_val = u32::from_le_bytes(image.data[offset..4].try_into()
+                                              .unwrap()
+            );
+            assert_eq!(input_val, read_val);
+        };
+
+        check_u32(manifest::ROM_EXT_MANIFEST_IDENTIFIER_OFFSET as usize,
```
Comment (mcy) [inline sw/host/rom_ext_image_tools/signer/image/src/image.rs:282]: This function is only used once...?

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 3]
```diff
@@ -184,19 +184,102 @@ impl RawImage {
...
 
     if value <= (u32::MAX as u64) {
         (value as u32).to_le_bytes().to_vec()
     } else {
         value.to_le_bytes().to_vec()
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (mcy) [inline sw/host/rom_ext_image_tools/signer/image/src/image.rs:209]: I think you need to run rustfmt?

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 4]
```diff
@@ -184,19 +184,102 @@ impl RawImage {
...
+use super::*;
+
+use std::convert::TryInto;
+
+use rom_ext_config::parser::Extension;
+use rom_ext_config::parser::InputFiles;
+
+use std::env;
```
Comment (mcy) [inline sw/host/rom_ext_image_tools/signer/image/src/image.rs:217]: Looks like this wandered away from the `use std` block above.

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs (hunk 2 above)
Comment (silvestrst) [reply sw/host/rom_ext_image_tools/signer/image/src/image.rs:282]: Oh, I forgot to add other asserts! Thanks!

Comment (silvestrst) [reply sw/host/rom_ext_image_tools/signer/image/src/image.rs:282]: Done.

**Review by silvestrst — COMMENTED** [review]
@mcy thanks a lot for the review!

- I have resolved all of your comments
- Added another minor commit (split out `update_time_stamp` into own public function for reasons described in the respective commit message)
- Testing of all fields is done in a single loop per byte basis. In case the byte falls into one of the fields that is tested by this test, it makes sure that the value is correct, all other bytes in the image must be zero.

@mcy could you please suggest a more ergonomic way of writing the loop at the end of the test function? I tried different approaches, but while loop was the only one that worked for me.

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 5]
```diff
@@ -184,19 +184,102 @@ impl RawImage {
...
-            .expect("Failed to parse string to u64!"),
-        false => s.parse::<u64>().expect("Failed to parse string to u64!"),
-    };
+    let value = str_to_u64(s);
 
     if value <= (u32::MAX as u64) {
         (value as u32).to_le_bytes().to_vec()
     } else {
         value.to_le_bytes().to_vec()
     }
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/host/rom_ext_image_tools/signer/image/src/image.rs]: Wait, then why isn't this just `value.to_le_bytes().to_vec()`?

**Review by silvestrst — COMMENTED** [review]
@mcy thanks again for the review, I have addressed the comments.

I have also realized that usage-constraints are 32-byte (256-bit) and not 256-byte long, so I changed that as well.

File: sw/host/rom_ext_image_tools/signer/image/src/image.rs [hunk 6]
```diff
@@ -90,10 +90,19 @@ impl RawImage {
...
+    /// integer, however there is no risk of overflow into a "negative".
+    pub fn update_timestamp_field(&mut self) {
+        let duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Failed to obtain the current time!");
 
-        // TODO: calculated at runtime, so we should probably rename this
-        //       function.
-        self.update_timestamp_field();
+        let bytes = &duration.as_secs().to_le_bytes();
```
[the remark this answers is not in the document]
Comment (silvestrst) [reply sw/host/rom_ext_image_tools/signer/image/src/image.rs:104]: Done.

**Comment by silvestrst** [comment]
@mcy - I think I've addressed all your feedback, can you please take a look and let me know if this is good to land from your perspective?

**Review by mcy — APPROVED** [review]
LG

**Pull request merged**